build-tools/packages/build-cli/src/codeCoverage/getCoverageMetrics.ts

```diff
@@ -8,46 +8,135 @@ import { Parser } from "xml2js";
 import type { CommandLogger } from "../logging.js";
 
 /**
- * The type for the coverage report, containing the line coverage and branch coverage(in percentage)
+ * The type for the coverage metric, containing the line, method and branch coverage(in percentage)
  */
 export interface CoverageMetric {
 	lineCoverage: number;
 	branchCoverage: number;
+	methodCoverage: number;
+}
+
+/**
+ * The type for the coverage metric for files in package, containing the line, method and branch coverage(in percentage)
+ */
+export interface CoverageMetricForPackages extends CoverageMetric {
+	filesCoverage: Map<string, CoverageMetric>;
 }
 
 interface XmlCoverageReportSchema {
 	coverage: {
-		packages: XmlCoverageReportSchemaForPackage[];
+		"$": {
+			"line-rate": string;
+			"branch-rate": string;
+		};
+		packages: [
+			{
+				package: XmlCoverageReportSchemaForPackage[];
+			},
+		];
 	};
 }
 
 interface XmlCoverageReportSchemaForPackage {
-	package: [
+	"$": {
+		name: string;
+		"line-rate": string;
+		"branch-rate": string;
+	};
+	classes: [
+		{
+			class?: XmlCoverageReportSchemaForFile[];
+		},
+	];
+}
+
+interface XmlCoverageReportSchemaForFile {
+	"$": {
+		name: string;
+		"line-rate": string;
+		"branch-rate": string;
+		filename: string;
+	};
+	methods: [
 		{
-			"$": {
-				name: string;
-				"line-rate": string;
-				"branch-rate": string;
-			};
+			method?: XmlCoverageReportSchemaForMethod[];
 		},
 	];
 }
 
+interface XmlCoverageReportSchemaForMethod {
+	"$": {
+		name: string;
+		"hits": string;
+	};
+}
+
+function extractCoverageForEachFiles(
+	coverageForFiles: XmlCoverageReportSchemaForFile[] | undefined,
+): {
+	filesCoverage: Map<string, CoverageMetric>;
+	packageMethodCoverage: number;
+} {
+	const filesCoverage: Map<
+		string,
+		{
+			lineCoverage: number;
+			branchCoverage: number;
+			methodCoverage: number;
+		}
+	> = new Map();
+	if (coverageForFiles === undefined) {
+		return { filesCoverage, packageMethodCoverage: 100 };
+	}
```

Comment:
Reporting 100% method coverage when we have no files seems incorrect. Should we report `undefined` or something similar?